Chromosomal Instability (PR 3 of 3): The notebook with plots (#419)
AlexsLemonade/OpenPBTA-analysis: analyses/chromosomal-instability/chromosomal-instability.Rmd

---
title: "Chromosomal Instability: Breakpoints"
output:   
  html_notebook: 
    toc: true
    toc_float: true
author: Candace Savonen for ALSF - CCDL
date: 2020
---

This analysis evaluates chromosomal instability by formatting SV and CNV data 
into chromosomal breaks and measuring chromosomal break density.
This notebook returns plots for each sample and tumor type group in the 
`plots` directory.

You can see the output notebook [here](https://cansavvy.github.io/openpbta-notebook-concept/chromosomal-instability/chromosomal-instability.nb.html).

Code adapted from [svcnvplus](https://github.com/gonzolgarcia/svcnvplus).

### Usage

This notebook can be run via the command line from the top directory of the 
repository as follows:

```
Rscript -e "rmarkdown::render('analyses/chromosomal-instability/chromosomal-instability.Rmd', 
                              clean = TRUE)"
```

### Set Up

```{r}

> # This threshold will be used to determine percent of genome changed

Author: This should be `0.2` so it would be 20%. I accidentally added a `0`. 😬

hist_plots_dir <- file.path(plots_dir, "tumor-type")

# Create the hist_plots_dir  if it does not exist
if (!dir.exists(hist_plots_dir)) {
  dir.create(hist_plots_dir, recursive = TRUE)
}

# Path to tumor type plots output directory
sample_plots_dir <- file.path(plots_dir, "sample")

# Create the sample_plots_dir if it does not exist
if (!dir.exists(sample_plots_dir)) {
  dir.create(sample_plots_dir, recursive = TRUE)
}
```

### Read in data and format it 

#### Set up metadata

```{r}
# Read in the metadata
metadata <- readr::read_tsv(file.path(data_dir, "pbta-histologies.tsv"))
```

#### Set up CNV data: 

Read in the CNV data and format it to make the chromosome and coordinates easier for
handling later. 

```{r}
# Read in the segment copy number data
cnv_df <- data.table::fread(file.path(data_dir, "pbta-cnv-cnvkit.seg.gz"),
  data.table = FALSE
) %>%
  dplyr::mutate(
    samples = as.factor(ID),
    # Calculate width
    width = loc.end - loc.start,
    # Make a binary variable that labels whether or not a sequence
    # is considered changed by the threshold set
    changed = as.factor(dplyr::case_when(
      seg.mean < log2(1 - ch.pct) ~ "change",

>       seg.mean >= log2(1 - ch.pct) ~ "no_change"

Author: This is not exactly right. Good call.

    ))
  ) %>%
  # Reformat the chromosome variable to drop the "chr"
  dplyr::mutate(chrom = factor(gsub("chr", "", chrom),
    levels = c(1:22, "X", "Y")
  )) %>%
  # Changing these so they end up matching the SV data
  dplyr::rename(start = loc.start, end = loc.end)
```

> Filter CNV data to only the changes that are larger than our cutoff `ch.pct`.

Author: I am unsure. Probably less crucial, yes.

```{r}
cnv_filtered_df <- cnv_df %>%
  dplyr::filter(changed == "change")
```

#### Set up SV data

Read in the SV data.

```{r}
sv_df <- data.table::fread(file.path(data_dir, "pbta-sv-manta.tsv.gz"), 
                           data.table = FALSE) %>%
  # Reformat the 23 and 24 chromosomes so they are X and Y and also factors
  dplyr::mutate(
    chrom = factor(dplyr::recode(SV.chrom,
      `23` = "X", `24` = "Y"
    ),
    levels = c(1:22, "X", "Y")
    )
  )
```

#### Set up chromosome reference data

Set up chromosome size data 

```{r}
# Set up Chr sizes

> chr_sizes <- readr::read_tsv(file.path(data_dir, "WGS.hg38.strelka2.unpadded.bed"),

Author: We need chromosome sizes based on the hg38 genome and this is what this file already happens to have. I'm all for getting the chr sizes in a less confusing/more elegant way though if you have a better alternative.

> chr_sizes <- readr::read_tsv(file.path(data_dir, "WGS.hg38.strelka2.unpadded.bed"),

Author: I haven't changed this. Do you have a suggestion for an alternative way to get chromosome sizes, @jashapiro ?

> chr_sizes <- readr::read_tsv(file.path(data_dir, "WGS.hg38.strelka2.unpadded.bed"),

Reviewer: Not really. But maybe just note that this file, despite its strange name, really does list only full chromosomes.

  col_names = FALSE
) %>%
  # Reformat the chromosome variable to drop the "chr"
  dplyr::mutate(X1 = factor(gsub("chr", "", X1),
    levels = c(1:22, "X", "Y", "M")
  )) %>%
  # Remove sex chromosomes
  dplyr::filter(!(X1 %in% c("X", "Y", "M")))

# Make chromosome size list

> chr_sizes_list <- chr_sizes$X3

Author: It was originally before I changed it to not be that way. The comment and object naming is outdated.

names(chr_sizes_list) <- chr_sizes$X1
```

## Format the data as chromosomes breaks

The SV and CNV data comes to us in the form of ranges, but for getting a look 
at chromosomal instability, we will want to convert this data into single break
points of the genome. 

> We'll also remove the sex chromosomes. 

Author: I am unsure what is common practice in dealing with CNVs on sex chromosomes and the data look a bit crazy based on what I've seen elsewhere.

Make an CNV breaks data.frame. 

```{r}
# Make a CNV data.frame that has the breaks
cnv_breaks <- data.frame(
  samples = rep(cnv_filtered_df$ID, 2),

>   chrom = c(cnv_filtered_df$chrom, cnv_filtered_df$chrom),

Author: No reason, just missed it.

1) `common_density` contains the combined break counts for both SV and CNV break data.   
2) `cnv_density` contains the number of break counts for CNV.   
3) `sv_density` contains the number of break counts for SV.  

```{r, warning = FALSE, results='hide'}
# Change here and it will change the rest
bin_size <- 1e6

# Obtain a list of samples that are common to both datasets. 
common_samples <- intersect(sv_breaks$samples, cnv_breaks$samples)

# Get a big list of break densities for each sample.
sample_densities <- lapply(common_samples, function(sample_id) {

  # Calculate break densities with both datasets combined
  common_density <- break_density(
    sv_breaks,
    cnv_breaks,
    sample_id = sample_id,
    window_size = bin_size,
    chr_sizes_list = chr_sizes_list
  )

  # Calculate break densities for CNV only
  cnv_density <- break_density(
    cnv_breaks,
    sample_id = sample_id,
    window_size = bin_size,
    chr_sizes_list = chr_sizes_list
  )

  # Calculate break densities for SV only
  sv_density <- break_density(
    sv_breaks,
    sample_id = sample_id,
    window_size = bin_size,
    chr_sizes_list = chr_sizes_list

>   )

Author: Okay. I willi streamline this section a bit. This is something I had had a concern/question about but wasn't sure what the best way to format it was.

  # Return this mega GRanges list
  return(list(
    common_density = common_density,
    cnv_density = cnv_density,
    sv_density = sv_density
  ))
})

# Bring along the sample IDs
names(sample_densities) <- common_samples
```

### Plot the breakpoints for each sample

Given the `GenomicRanges` objects for each sample, create a combined plot for 
each. 

```{r, results='hide'}
lapply(sample_densities, function(granges_list) {

  # Get the sample name

>   sample_name <- common_samples[parent.frame()$i[]]

Author: Cool. I'll try out something like this.

  # Make the combo plot
  multipanel_break_plot(
    granges_list = granges_list,
    plot_name = sample_name,
    y_val = "total_counts",
    y_lab = "Breaks per Mb",
    plot_dir = sample_plots_dir
  )
})
```

### Co-localization of breakpoints for tumor-type groups

Same as was done for each sample, now we will calculate densities for 
each tumor-type group. 

```{r, results='hide'}
# Change bin_size here and it will change the rest

> bin_size <- 1e6

Author: No, just wanted it to be clear that this next set of plots will also be using a bin_size so I copied it so we could remind ourselves of what it is and/or change it if needed.

# Get a list of the tumor_types for which we have DNA-seq data
tumor_types <- metadata %>% 
  dplyr::filter(!is.na(short_histology), experimental_strategy != "RNA-Seq") %>% 
  dplyr::distinct(short_histology) %>%
  dplyr::pull(short_histology)

> tumor_types[12]

Author: This wasn't supposed to be here and is a remnant from my troubleshooting.

    sample_id = sample_ids,
    window_size = bin_size,
    chr_sizes_list = chr_sizes_list
  )

  # Calculate break densities for SV only
  sv_density <- break_density(
    sv_breaks,
    sample_id = sample_ids,
    window_size = bin_size,
    chr_sizes_list = chr_sizes_list
  )

  # Return this mega GRanges list
  return(list(
    common_density = common_density,
    cnv_density = cnv_density,
    sv_density = sv_density
  ))
  }
})

# Bring along the tumor-type labels
names(group_densities) <- tumor_types

# Remove tumor_types data that didn't have at least two samples
group_densities <- group_densities[!sapply(group_densities, is.null)]
```

### Plot the breakpoints for each tumor-type

Here we will plot median number of break points for the tumor-type group per 
each bin.

```{r}
lapply(group_densities, function(granges_list) {
  # Get the tumor type label
  group_name <- tumor_types[parent.frame()$i[]]

  # Make the combo plot
  multipanel_break_plot(
    granges_list = granges_list,
    plot_name = group_name,
    y_val = "median_counts",
    y_lab = "Median Breaks per Mb",
    plot_dir = hist_plots_dir
  )
})
```

Zip up the PNG files into one file. 

```{r}

> zip(paste0(hist_plots_dir, ".zip"), 

Reviewer: somehow this ended up with a folder within a folder... My guess is this happens if the zip file already exists from a previous run.

Author: This should be fixed with the most recent commit I did.

Set up chromosome size data. 
It just so happens that this BED file: `WGS.hg38.strelka2.unpadded.bed` is actually 
just a list of the chromosome sizes so we are using that for now. 

```{r}
# Set up Chr sizes
chr_sizes <- readr::read_tsv(file.path(data_dir, "WGS.hg38.strelka2.unpadded.bed"),
  col_names = FALSE
) %>%
  # Reformat the chromosome variable to drop the "chr"
  dplyr::mutate(X1 = factor(gsub("chr", "", X1),
    levels = c(1:22, "X", "Y", "M")
  )) %>%
  # Remove sex chromosomes
  dplyr::filter(!(X1 %in% c("X", "Y", "M")))

# Make chromosome size named vector
chr_sizes_vector <- chr_sizes$X3
names(chr_sizes_vector) <- chr_sizes$X1
```

## Format the data as chromosomes breaks

The SV and CNV data comes to us in the form of ranges, but for getting a look 
at chromosomal instability, we will want to convert this data into single break
points of the genome. 
We'll also remove the sex chromosomes. 

Make an CNV breaks data.frame. 

```{r}
# Make a CNV data.frame that has the breaks
cnv_breaks <- data.frame(
  samples = rep(cnv_filtered_df$ID, 2),
  chrom = rep(cnv_filtered_df$chrom, 2),
  start = c(cnv_filtered_df$start, cnv_filtered_df$end),

>   end = c(cnv_filtered_df$start, cnv_filtered_df$end),

Author: `GenomicRanges` wants a start and an end even if they are the same.

>   end = c(cnv_filtered_df$start, cnv_filtered_df$end),

Reviewer: I think I realized that later. But then I would do it when making the genomic ranges, not here.

---
title: "Chromosomal Instability: Breakpoints"
output:   
  html_notebook: 
    toc: true
    toc_float: true
author: Candace Savonen for ALSF - CCDL
date: 2020
---

This analysis evaluates chromosomal instability by formatting SV and CNV data 
into chromosomal breaks and measuring chromosomal break density.
This notebook returns chromosomal break heatmaps and plots for each tumor type 
group in the `plots` directory.

Code adapted from [svcnvplus](https://github.com/gonzolgarcia/svcnvplus).

### Usage

This notebook can be run via the command line from the top directory of the 
repository as follows:

```
Rscript -e "rmarkdown::render('analyses/chromosomal-instability/chromosomal-instability.Rmd', 
                              clean = TRUE)"
```

### Set Up

```{r}
# Set seed so heatmaps turn out the same
set.seed(2020)

> # This threshold will be used to determine percent of genome changed

Reviewer: ```suggestion
# This threshold will be used to determine the threshold copy number changes identified.
```

at chromosomal instability, we will want to convert this data into single break
points of the genome. 
We'll also remove the sex chromosomes. 
TODO: after updating to consensus CNV data, evaluate whether sex chromosomes should still be removed. 

Make an CNV breaks data.frame. 

```{r}
# Make a CNV data.frame that has the breaks
cnv_breaks <- data.frame(
  samples = rep(cnv_filtered_df$ID, 2),
  chrom = rep(cnv_filtered_df$chrom, 2),
  coord = c(cnv_filtered_df$start, cnv_filtered_df$end),
  seg.mean = rep(cnv_filtered_df$seg.mean, 2),
  copy.num = rep(cnv_filtered_df$copy.num, 2),
  stringsAsFactors = FALSE
) %>%
  # Remove sex chromosomes
  dplyr::filter(!(chrom %in% c("X", "Y")))
```

Make an SV breaks data.frame. 

```{r}
# Make a data.frame that has the breaks
sv_breaks <- data.frame(
  samples = rep(sv_df$Kids.First.Biospecimen.ID.Tumor, 2),
  chrom = rep(sv_df$chrom, 2),
  coord = c(sv_df$SV.start, sv_df$SV.end),
  svclass = rep(sv_df$SV.type, 2),
  stringsAsFactors = FALSE
) %>%
  # Remove sex chromosomes and NAs
  dplyr::filter(!(chrom %in% c("X", "Y", "M", NA)))
```

Make an common breaks data.frame. 

```{r}
common_breaks <- dplyr::bind_rows(sv_breaks, cnv_breaks) %>%

>   dplyr::distinct(samples, chrom, coord, .keep_all = TRUE)

Reviewer: There was some discussion as to whether this should be an intersection, rather than union (which this is), but this seems correct to me. The only question I had, which we did discuss verbally, was whether we should also filter out adjacent breakpoints, where one segment starts 1 bp after the previous segment. I don't anticipate this will happen much in the consensus data, but it is a thing that could occur, and would overcount high instability regions in particular.
